geomesa-core/src/main/scala/org/locationtech/geomesa/core/csv/package.scala

```diff
@@ -39,84 +39,80 @@ import scala.collection.JavaConversions._
-        case Some(x) => x._2
-        case None    => 's'   // should get to this anyway as StringParser is guaranteed to succeed
-      }
+  def csvToFeatures(csvFile: File,
+                    hasHeader: Boolean,
+                    typeSchema: TypeSchema): SimpleFeatureCollection = {
+    val sft = SimpleFeatureTypes.createType(typeSchema.name, typeSchema.schema)
+    buildFeatureCollection(csvFile, hasHeader, sft, typeSchema.latLonFields)
+  }
 
+  protected[csv] def tryParsers(rawData: TraversableOnce[String]): Seq[CSVParser[_]] = {
+    def tryAllParsers(datum: String) =
+      CSVParser.parsers.find(_.parse(datum).isSuccess).getOrElse(StringParser)
     rawData.map(tryAllParsers).toSeq
   }
-  
-  def sampleRecords(records: jIterator[CSVRecord], hasHeader: Boolean): (Seq[String], CSVRecord) =
-    if (hasHeader) {
-      val header = records.next
-      val record = records.next
-      (header.toSeq, record)
-    } else {
-      val record = records.next
-      val header = Seq.tabulate(record.size()) { n => s"C$n" }
-      (header, record)
-    }
 
-  def guessTypes(name: String,
+  protected[csv] def guessHeaders(record: CSVRecord, hasHeader: Boolean = true): Seq[String] =
+    if (hasHeader) record.toSeq else Seq.tabulate(record.size())(n => s"C$n")
+
+  protected[csv] def guessTypes(name: String,
                  csvReader: Reader,
                  hasHeader: Boolean = true,
-                 format: CSVFormat = CSVFormat.DEFAULT): TypeSchema = {
-    val records = format.parse(csvReader).iterator
-    val (header, record) = sampleRecords(records, hasHeader)
-    val typeChars = typeData(record.iterator)
+                 format: CSVFormat = CSVFormat.DEFAULT,
+                 numSamples: Int = 5): TypeSchema = {
+    assert(numSamples > 0)
+    val records = format.parse(csvReader).iterator.take(numSamples + 1).toSeq
+    assert(records.size > 1 || (!hasHeader && records.size > 0))
+
+    val headers = guessHeaders(records(0), hasHeader)
+    val sample = records.drop(1)
+
+    // make sure type chars are valid for first few records
+    val parsers = sample.map(tryParsers(_)).reduceLeft { (pc1, pc2)  =>
+      pc1.zip(pc2).map { case (p1, p2) => if (p1 == p2) p1 else StringParser }
+    }
 
     val sftb = new SftBuilder
-    var defaultDateSet = false
     var defaultGeomSet = false
-    for ((field, c) <- header.zip(typeChars)) { c match {
-      case 'i' =>
-        sftb.intType(field)
-      case 'd' =>
-        sftb.doubleType(field)
-      case 't' =>
-        sftb.date(field)
-        if (!defaultDateSet) {
-          sftb.withDefaultDtg(field)
-          defaultDateSet = true
-        }
-      case 'p' =>
-        if (defaultGeomSet) sftb.geometry(field)
-        else {
-          sftb.point(field, default = true)
-          defaultGeomSet = true
-        }
-      case 's' =>
-        sftb.stringType(field)
-    }}
+    headers.zip(parsers).foreach { case (field, parser) =>
```

Comment:
I suppose this works, because each of independent parser objects is declared inside the ```CSVParser``` object that gives them all access to the same shared builder, right?  This strikes me as somewhat less intuitive than the old approach in which the ```SftBuilder``` was declared locally.